0000/README.adoc

```diff
@@ -0,0 +1,175 @@
+transformation on the payload before value is injected. We define two trim strategies: `TO_NULL` (default one)
+and `TO_EMPTY`.
+
+== Motivation
+
+http://stapler.kohsuke.org/[Stapler] framework is the backbone of Jenkins web UI. It is designed to
+offer data-binding from a (jelly or groovy) web view to Java model. For this purpose, a JSON payload
+is sent when a configuration form is submitted, and Stapler do convert this back to java model classes
+using a "DataBound" constructor and optional setters.
+
+This mechanism offers some discoverable data model on the Java API : the DataBound constructor and setters
+let us know about the name and types of the expected payload. This is the basis for Configuration-as-Code
+(JEP-201). But this model is pretty poor: the only constraint we know on attribute is the expected type.
+In many cases, some additional requirement are checked with human-written code. Typically, parameters get
+trimmed, checked for being empty or null, ensure value belongs to some range or match a pattern, etc.
+
+On the other way, the web UI form is build by accessing live Java model data using read methods, being
+converted to plain strings. It's up to the plugin developer to define constraints on the html elements
+so they reflect the data model.
+
+Generally speaking, with every single line of human-written code to handle data binding we introduce
+development effort / risk to introduce bugs, an we block opportunity for third-party component to discover
+the underlying data model.
+
+
+== Reasoning
+
+Annotations have been used intensively by many frameworks to define metadata on data model expressed as
+Java classes. Typically, https://jcp.org/en/jsr/detail?id=338[Java Persistence API] allows to define
+how a Java object can be mapped to a database, while https://jcp.org/en/jsr/detail?id=222[JAXB] define
+annotations to customize how java attribute will be mapped to xml elements.
+
+This JEP is about introducing support for annotation-based declarative data binding. A DataBound jenkins
+component would only rely on annotation to define how attributes are bound to JSON payload from web forms,
+including validation constraints. We consider this would have two major benefits :
+
+1. Plugin developers will need to write less code, and use standard annotations they probably already know
+from other Java related developments.
+
+2. Constraints on the data model can be introspected. This will allow third party tools to adjust their
+behavior accordingly. A good candidates for this is Configuration-as-Code (JEP-201). Jelly tags use to
+compose the web UI also could introspect those annotations. Typically, if some attribute is a number and
+has to be positive, `<input type='number' 'min=0'>` html 5 element would be automatically generated.
+
+==== Relying on Bean Validation
+
+https://jcp.org/en/jsr/detail?id=222[Bean Validation] is an annotation-based framework designed for
```

Comment:
[This](http://beanvalidation.org/2.0/spec/) seems like a more useful link. I had to Google around quite a bit to find the actual documentation.